packages/react-slider/src/components/Slider/useSliderState.tsx

```diff
@@ -0,0 +1,236 @@
+      const size = currentBounds?.width || 0;
+      const position = currentBounds?.left || 0;
+
+      const totalSteps = (max - min) / step;
+      const stepLength = size / totalSteps;
+      const thumbPosition = ev.clientX;
+      const distance = thumbPosition - position;
+      return distance / stepLength;
+    },
+    [max, min, step],
+  );
+
+  const onPointerMove = React.useCallback(
+    (ev): void => {
+      if (snapToStep || step !== 1) {
+        updateValue(ev, Math.round((min + step * calculateSteps(ev)) / step) * step);
+      } else {
+        updateValue(ev, min + step * calculateSteps(ev));
+      }
+    },
+    [calculateSteps, min, snapToStep, step, updateValue],
+  );
+
+  const onPointerUp = (): void => {
+    disposables.current.forEach(dispose => dispose());
+    disposables.current = [];
+  };
+
+  const onPointerDown = React.useCallback(
+    (ev): void => {
+      const { currentTarget, pointerId } = ev;
+
+      onPointerDownCallback?.(ev);
+
+      if (currentTarget.setPointerCapture) {
+        currentTarget.setPointerCapture(pointerId);
+      }
+
+      disposables.current.push(
+        on(currentTarget, 'pointermove', onPointerMove),
+        on(currentTarget, 'pointerup', onPointerUp),
+        () => {
+          currentTarget.releasePointerCapture(pointerId);
+        },
+      );
+
+      onPointerMove(ev);
+    },
+    [onPointerMove, onPointerDownCallback],
+  );
+
+  const onKeyDown = React.useCallback(
+    (ev: React.KeyboardEvent<HTMLDivElement>): void => {
+      const key = getCode(ev);
+
+      onKeyDownCallback?.(ev);
+
+      if (ev.shiftKey) {
+        if (key === ArrowDownKey || key === ArrowLeftKey) {
+          updateValue(ev, currentValue! - step * 10);
+          return;
+        } else if (key === ArrowUpKey || key === ArrowRightKey) {
+          updateValue(ev, currentValue! + step * 10);
+          return;
+        }
+      } else if (key === ArrowDownKey || key === ArrowLeftKey) {
+        updateValue(ev, currentValue! - step);
+        return;
+      } else if (key === ArrowUpKey || key === ArrowRightKey) {
+        updateValue(ev, currentValue! + step);
+        return;
+      } else {
+        switch (key) {
+          case PageDownKey:
+            updateValue(ev, currentValue! - step * 10);
+            break;
+          case PageUpKey:
+            updateValue(ev, currentValue! + step * 10);
+            break;
+          case HomeKey:
+            updateValue(ev, min);
+            break;
+          case EndKey:
+            updateValue(ev, max);
+            break;
+        }
+      }
+    },
+    [currentValue, max, min, onKeyDownCallback, step, updateValue],
+  );
+
+  React.useEffect(() => {
+    if (state.ref && state.ref.current) {
+      state.ref.current.value = currentValue;
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      state.ref.current.focus = thumbRef?.current?.focus() as any;
+    }
+  }, [currentValue, state.ref]);
```

Comment:
This might not be required once #18983 will be implemented? _As `ref` will be passed to `input` element._